aten/src/ATen/native/cuda/ForeachBinaryOpList.cu

```diff
@@ -278,16 +412,38 @@ void foreach_tensor_copy_list_kernel_cuda_(
       "foreach_tensor_copy",
       [&]() {
         using opmath_t = at::opmath_type<scalar_t>;
-        multi_tensor_apply<2>(
-            tensor_lists,
-            UnaryOpFunctor<
-                scalar_t,
-                /* depth */ 2,
-                /* r_args_depth */ 1,
-                /* res_arg_index */ 1>(),
-            Identity<opmath_t>());
+        AT_DISPATCH_SOURCE_TYPES(src[0].scalar_type(), "foreach_tensor_copy", [&] {
+          if constexpr (std::is_same_v<scalar_t, src_t>) {
+            multi_tensor_apply<2>(
+                tensor_lists,
+                UnaryOpFunctor<
+                    scalar_t,
+                    /* depth */ 2,
+                    /* r_args_depth */ 1,
+                    /* res_arg_index */ 1>(),
+                Copy<opmath_t, opmath_t>());
+          } else {
+            // Ref:
+            // https://github.com/pytorch/pytorch/blob/656134c38f4737d13c3f43fc5c59470bc23c1d2f/aten/src/ATen/native/Copy.cpp#L299-L301
+            if (!self[0].is_complex() && src[0].is_complex()) {
+              TORCH_WARN_ONCE(
+                  "Casting complex values to real discards the imaginary part");
+            }
+            multi_tensor_apply<2>(
+                tensor_lists,
+                CopyFunctor<
```

Comment:
Another q for my understanding--what is the difference between the CopyFunctor and a normal UnaryFunctor? I'd expect the structure between the "diff dtype" and "same dtype" implementations to be similar, just one has a cast.

Reply:
`CopyFunctor` uses different types for self and src while in UnaryFunctor they have the same type. Also, `CopyFunctor` does not cast inputs to a given functor to `opmath_t`

Reply:
Ah, but if we relaxed the UnaryFunctor signature to allow different types, then it could be reused?

Reply:
yes but I thought it's too much until we have multiple use cases